### PR TITLE
slightly refactor test_expect for operators to use correct variance

### PR DESCRIPTION
**Pull request by PhilipVinc**
This PR changes the logic used to compute the distance between two gradients (i believe the old one was partly broken).
Also the error tolerance now depends on the variance of the local energy. 

Taken from #1065 . I want to merge the changes to the tests separately and be sure you agree with them.

### CODE REVIEW

**Comment by femtobit** [comment]
>  Also the error tolerance now depends on the variance of the local energy.

The MC standard error scales with `mcse = sqrt(variance / N)`. Is there a reason the tolerances are now given in terms of `err = variance / sqrt(N)`? (Instead of, say, a factor times `mcse`?)

**Comment by PhilipVinc** [comment]
ah right, it's just my fault.

```python
err =  np.sqrt (O_stat.variance / vstate.n_samples)
```

The reason i want this instead of a constant is that when estimating the error on Squared operators (whose variance is larger) the one with `variance=5` that is now used fails.

**Comment by femtobit** [comment]
Its certainly better to check whether the expectation value falls approximately into `true_value ± mcse`.

But you probably need a factor: Assuming everything has a nice Gaussian distribution, the estimate would end up outside the standard error bounds about a third of the time, right?

File: test/variational/test_variational.py [hunk 1]
```diff
@@ -405,7 +408,7 @@ def expval_fun(par, vstate, H):
         grad_exact = jax.tree_map(lambda x: x * 2, grad_exact)
 
     # compare the two
-    err = 5 / np.sqrt(vstate.n_samples)
```
Comment (gcarleo) [inline test/variational/test_variational.py:408]: that's why here I put a 5 I think

**Review by gcarleo — CHANGES_REQUESTED** [review]

File: test/variational/test_variational.py [hunk 2]
```diff
@@ -371,15 +372,17 @@ def test_expect(vstate, operator):
 
     O1_mean = np.asarray(O_stat1.mean)
     O_mean = np.asarray(O_stat.mean)
+    err = O_stat1.variance / np.sqrt(vstate.n_samples)
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply test/variational/test_variational.py]: i forgot we precompute that

File: test/variational/test_variational.py [hunk 3]
```diff
@@ -500,9 +503,20 @@ def same_derivatives(der_log, num_der_log, abs_eps=1.0e-6, rel_eps=1.0e-6):
     np.testing.assert_allclose(
         der_log.real, num_der_log.real, rtol=rel_eps, atol=abs_eps
     )
+
+    # compute the distance between the two phases modulo 2pi
+    der_log_mod = np.mod(der_log.imag, np.pi * 2)
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply test/variational/test_variational.py]: Yeah I can, i was just being overly zealous...

I simplified a bit this logic now

**Review by femtobit — APPROVED** [review]
Looks good to me now.

File: test/variational/test_variational.py [hunk 4]
```diff
@@ -500,9 +501,20 @@ def same_derivatives(der_log, num_der_log, abs_eps=1.0e-6, rel_eps=1.0e-6):
     np.testing.assert_allclose(
         der_log.real, num_der_log.real, rtol=rel_eps, atol=abs_eps
     )
+
+    # compute the distance between the two phases modulo 2pi
+    delta_der_log = der_log.imag - num_der_log.imag
+
+    # the distance is taken to be the minimum of the distance between |A-B| and
+    # mod(|B-A|)
```
Comment (gcarleo) [inline test/variational/test_variational.py:509]:
```suggestion
 # |B-A| mod 2Pi
```

**Review by gcarleo — APPROVED** [review]
looks good now
